lib/Dialect/FIRRTL/Transforms/GrandCentral.cpp

```diff
@@ -0,0 +1,378 @@
+//===- GrandCentral.cpp - Ingest black box sources --------------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//===----------------------------------------------------------------------===//
+//
+// Implement SiFive's Grand Central transform.
+//
+//===----------------------------------------------------------------------===//
+
+#include "./PassDetails.h"
+#include "circt/Dialect/FIRRTL/FIRRTLVisitors.h"
+#include "circt/Dialect/FIRRTL/Passes.h"
+#include "circt/Dialect/SV/SVOps.h"
+#include "mlir/IR/ImplicitLocOpBuilder.h"
+
+using namespace circt;
+using namespace firrtl;
+
+//===----------------------------------------------------------------------===//
+// Pass Implementation
+//===----------------------------------------------------------------------===//
+
+namespace {
+struct Options {
+  StringAttr extractGCTAnno;
+  StringAttr augmentedBundleType;
+  StringAttr augmentedGroundType;
+  StringAttr augmentedVectorType;
+  StringAttr serializedViewAnno;
+  StringAttr directory;
+  StringAttr filename;
+};
+
+namespace Component {
+enum class Type { Unknown = -1, Ground, Vector, Bundle };
+
+struct Info {
+  Type tpe;
+  StringRef description;
+  bool found = false;
+  int32_t width = -1;
+  uint32_t depth = 0;
+};
+} // namespace Component
+
+struct GrandCentralPass : public GrandCentralBase<GrandCentralPass> {
+  void runOnOperation() override;
+
+private:
+  // Discovered interfaces that need to be constructed.
+  llvm::DenseMap<std::pair<StringRef, StringRef>, Component::Info> interfaceMap;
+
+  // Discovered views that need mappings.  Keyed by the view "defName".
+  llvm::StringMap<FModuleOp> mappingMap;
+};
+
+class GrandCentralVisitor : public FIRRTLVisitor<GrandCentralVisitor> {
+public:
+  GrandCentralVisitor(Options *opts, MLIRContext *context,
+                      llvm::DenseMap<std::pair<StringRef, StringRef>,
+                                     Component::Info> *interfaceMap)
+      : options(opts), context(context), interfaceMap(interfaceMap) {}
+
+  using FIRRTLVisitor<GrandCentralVisitor>::visitDecl;
+
+  void visitDecl(FModuleOp op);
+  void visitDecl(FExtModuleOp op);
+  void visitDecl(RegOp op);
+  void visitDecl(WireOp op);
+  void visitDecl(NodeOp op);
+  void visitUnhandledOp(Operation *op) {}
+
+  ArrayAttr getAnnotations(Operation *op) {
+    return op->getAttrOfType<ArrayAttr>("annotations");
```

Comment:
Please use `AnnotationSet` instead of poking at the array/dicts directly.

Reply:
Switched over to this throughout.